### PR TITLE
Adjust select timeouts during port update handling to allow for faster transceiver DOM polling

### LINKED ISSUE #759: DomInfoUpdateTask polling can take longer than DOM_INFO_UPDATE_PERIOD_SECS due to 1 sec wait after every iteration
`DomInfoUpdateTask` polling loop is structured as follows:
```
            for physical_port, logical_ports in self.port_mapping.physical_to_logical.items():
                # Process pending link change events and update diagnostic
                # information in the database. Ensures timely handling of link
                # change events and avoids duplicate updates in case of breakout ports.
                port_change_observer.handle_port_update_event() <<=============
                # Process each port in the pending link change set based on the
                # corresponding time to update the DB after the link change.
                for link_changed_port in list(self.link_change_affected_ports.keys()):
                    if self.task_stopping_event.is_set():
                        self.log_notice("Stop event generated during DOM link change event processing")
                        break
                    if self.link_change_affected_ports[link_changed_port] <= datetime.datetime.now():
                        self.log_notice(f"Updating port db diagnostics post link change for port {link_changed_port}")
                        self.update_port_db_diagnostics_on_link_change(link_changed_port)
                        del self.link_change_affected_ports[link_changed_port]

                if self.task_stopping_event.is_set():
                    self.log_notice("Stop event generated during DOM monitoring loop")
                    break

                if not is_periodic_db_update_needed:
                    # If periodic db update is not needed, skip the rest of the loop
                    continue

                # Get the first logical port name since it corresponds to the first subport
                # of the breakout group
                logical_port_name = logical_ports[0]
...
```

`port_change_observer.handle_port_update_event()` call goes to `xcvrd/xcvrd_utilities/port_event_helper.py`:
```
    def handle_port_update_event(self):
        """
        Select PORT update events, notify the observers upon a port update in CONFIG_DB
        or a XCVR insertion/removal in STATE_DB

        Returns:
            bool: True if there's at least one update event; False if there's no update event.
        """
        has_event = False
        if not self.stop_event.is_set():
            (state, _) = self.sel.select(SELECT_TIMEOUT_MSECS)
            if state == swsscommon.Select.TIMEOUT:
                return has_event
            if state != swsscommon.Select.OBJECT:
                self.logger.log_warning('sel.select() did not return swsscommon.Select.OBJECT')
                return has_event
...
```
Here, `SELECT_TIMEOUT_MSECS = 1000`

So when there are no port change events, we spend 1 sec waiting for an event in `handle_port_update_event`.
On systems with a large number of transceivers or with many VDM enabled transceivers (where polling takes longer), we are unable to poll information from all transceivers in DOM_INFO_UPDATE_PERIOD_SECS (currently 60 secs)

This leads to flags and DOM information updating slower than 60 seconds.

### PR DESCRIPTION
**Pull request by aditya-nexthop**
#### Description
This PR optimizes the DOM (Digital Optical Monitoring) polling loop in xcvrd by improving port update event handling and reducing unnecessary wait times. The changes include:

1. **Refactored port update handling** - Extracted port update event processing into a dedicated `check_port_update()` method for better code organization and reusability
2. **Optimized timeout strategy** - Introduced two-tier timeout mechanism:
   - 1000ms timeout when waiting for port updates before DOM polling begins and after when it completes until the next polling cycle
   - 100ms timeout during DOM polling to minimize delays while still handling port events
3. **Improved loop structure** - Restructured the main loop to handle port updates before entering the DOM polling phase, preventing unnecessary blocking
4. **Added comprehensive unit tests** - Created tests covering various scenarios including multiple ports, timing edge cases, and stop event handling

fixes #759 and builds on https://github.com/sonic-net/sonic-platform-daemons/pull/757

#### Motivation and Context
Before the change, the DOM monitoring loop would wait up to 1 second (`SELECT_TIMEOUT_MSECS`) for port update events during each iteration of the physical port loop. This caused significant delays in DOM data collection, especially on systems with many ports.

**Problem:** With the 1-second timeout being called for every physical port, the DOM polling could take an excessive amount of time to complete, delaying DOM polling updates.

**Solution:** By separating port update handling from DOM polling and using a shorter 100ms timeout during the polling phase, the loop can complete much faster while still being responsive to port change events. The 1-second timeout is only used when explicitly waiting for port updates before starting the next DOM polling cycle.

#### How Has This Been Tested?
1. **Unit Tests Added** - Comprehensive test coverage for the new `check_port_update()` method including:
   - Scenario with no link change affected ports
   - Link change affected port with past timestamp (should trigger immediate update)
   - Link change affected port with future timestamp (should defer update)
   - Multiple ports with mixed ready/not-ready states
   - Stop event handling during processing

2. **CPU usage profiling**: Measured for 10 minutes after restarting `xcvrd` on a switch fully populated with optical transceivers.
- Before the change (includes changes from https://github.com/sonic-net/sonic-platform-daemons/pull/757)
image
- After the change
image

The CPU usage is slightly higher during active polling as it spends less time waiting (100ms) between interfaces and then the loop spends time waiting for port change updates in 1s chunks.

3. **Measuring actual DOM update times**
- Before change: We are not able to poll every 60 seconds (takes 90 sec+) on a switch fully populated with optical transceivers
```
$ while true; do sonic-db-dump -n STATE_DB -y -k "TRANSCEIVER_DOM_SENSOR|Ethernet112" | grep last_update_time; sleep 10; done | uniq
      "last_update_time": "Tue Feb 24 19:25:57 2026",
      "last_update_time": "Tue Feb 24 19:27:32 2026",
      "last_update_time": "Tue Feb 24 19:29:06 2026",
      "last_update_time": "Tue Feb 24 19:30:40 2026",
      "last_update_time": "Tue Feb 24 19:32:13 2026",
      "last_update_time": "Tue Feb 24 19:33:47 2026",
```
- After change: Updates happen every 60 seconds for a specified interface
```
$ while true; do sonic-db-dump -n STATE_DB -y -k "TRANSCEIVER_DOM_SENSOR|Ethernet112" | grep last_update_time; sleep 10; done | uniq
      "last_update_time": "Tue Feb 24 19:17:09 2026",
      "last_update_time": "Tue Feb 24 19:18:10 2026",
      "last_update_time": "Tue Feb 24 19:19:10 2026",
      "last_update_time": "Tue Feb 24 19:20:10 2026",
      "last_update_time": "Tue Feb 24 19:21:11 2026",
      "last_update_time": "Tue Feb 24 19:22:11 2026",
```

#### Additional Information (Optional)
**Key Technical Changes:**
- New constants: `PORT_UPDATE_EVENT_SELECT_TIMEOUT_MSECS` (1000ms) and `PORT_UPDATE_EVENT_SELECT_TIMEOUT_FAST_MSECS` (100ms)
- Modified `PortChangeObserver.handle_port_update_event()` to accept a configurable timeout parameter
- The periodic update interval calculation now uses `dom_loop_start_time` to maintain consistent intervals regardless of loop execution time

**Backward Compatibility:** This change is fully backward compatible and does not affect the external API or configuration.

#### Tested branch
- [x] master
- [x] 202605: already in Nexthop's internal 202605 branch

### CODE REVIEW

File: sonic-xcvrd/xcvrd/dom/dom_mgr.py [hunk 1]
```diff
@@ -391,8 +397,7 @@ def task_worker(self):
 
             # Set the periodic db update time after all the ports are processed
             if is_periodic_db_update_needed:
-                next_periodic_db_update_time = datetime.datetime.now() + \
-                                               datetime.timedelta(seconds=dom_info_update_periodic_secs)
+                next_periodic_db_update_time = dom_loop_start_time + datetime.timedelta(seconds=dom_info_update_periodic_secs)
```
[the remark this answers is not in the document]
Comment (aditya-nexthop) [reply sonic-xcvrd/xcvrd/dom/dom_mgr.py]: We don't, as this PR includes the change from PR #757.
I posted them as 2 separate PRs so that we can clearly show the effect of both PRs independently.
If we merge this in directly, we can discard PR #757

File: sonic-xcvrd/xcvrd/dom/dom_mgr.py [hunk 2]
```diff
@@ -33,6 +33,8 @@
     raise ImportError(str(e) + " - required module not found in dom_mgr.py")
 
 SYSLOG_IDENTIFIER_DOMINFOUPDATETASK = "DomInfoUpdateTask"
+PORT_UPDATE_EVENT_SELECT_TIMEOUT_MSECS = 1000
+PORT_UPDATE_EVENT_SELECT_TIMEOUT_FAST_MSECS = 100
```
Comment (mihirpat1) [inline sonic-xcvrd/xcvrd/dom/dom_mgr.py]: @aditya-nexthop Can we reduce this to 10ms or a smaller value for better performance?

**Review by mihirpat1 — DISMISSED** [review]

**Review by prgeor — CHANGES_REQUESTED** [review]
@aditya-nexthop please address comment.

File: sonic-xcvrd/xcvrd/dom/dom_mgr.py [hunk 3]
```diff
@@ -271,37 +290,24 @@ def task_worker(self):
...
             # Handle port change event from main thread
             port_event_helper.handle_port_config_change(sel, asic_context, self.task_stopping_event, self.port_mapping, self.helper_logger, self.on_port_config_change)
 
+            while not (is_periodic_db_update_needed:= next_periodic_db_update_time <= datetime.datetime.now()):
+                self.check_port_update(port_change_observer, PORT_UPDATE_EVENT_SELECT_TIMEOUT_MSECS)
+
+                if self.task_stopping_event.is_set():
+                    self.log_notice("Stop event generated during DOM monitoring loop while checking port update")
+                    break
+
```
[the remark this answers is not in the document]
Comment (prgeor) [reply sonic-xcvrd/xcvrd/dom/dom_mgr.py]: @aditya-nexthop instead of having an outer wait of 1 sec and then inner wait loop of 100msec, why not have one wait loop of 100msec?

[the remark this answers is not in the document]
Comment (aditya-nexthop) [reply sonic-xcvrd/xcvrd/dom/dom_mgr.py:319]: This is a good question. However, now we have outer wait of 1 sec and inner wait of 10ms (changed from 100ms)

If we are waiting for multiple seconds if the DOM polling finishes early, we start CPU spinning every 10ms until the DOM polling interval is completed leading to an unnecessary CPU usage increase. A long wait (like 1 sec) outer loop prevents that.

File: sonic-xcvrd/xcvrd/dom/dom_mgr.py [hunk 4]
```diff
@@ -392,7 +405,7 @@ def task_worker(self):
 
             # Set the periodic db update time after all the ports are processed
             if is_periodic_db_update_needed:
-                next_periodic_db_update_time = now + datetime.timedelta(seconds=dom_info_update_periodic_secs)
+                next_periodic_db_update_time = dom_loop_start_time + datetime.timedelta(seconds=dom_info_update_periodic_secs)
```
[the remark this answers is not in the document]
Comment (aditya-nexthop) [reply sonic-xcvrd/xcvrd/dom/dom_mgr.py]: Hi Prince,

The code flow pertaining to the DOM polling time is as follows:

1. `Start task_worker`
2. Set `next_periodic_db_update_time` to `datetime.datetime.now()` + `dom_info_update_periodic_secs` seconds to allow for initialization, `is_periodic_db_update_needed = False`
3. Run the 1000 msec polling loop until `is_periodic_db_update_needed = next_periodic_db_update_time <= datetime.datetime.now()` and `is_periodic_db_update_needed == True`
4.  Set `dom_loop_start_time = datetime.datetime.now()` <<<< before the actual DOM polling starts
5. Run the 10 msec select + poll interface loop
6. At the end, set `next_periodic_db_update_time = dom_loop_start_time + dom_info_update_periodic_secs`

We then go back to 3. and wait until `is_periodic_db_update_needed = next_periodic_db_update_time <= datetime.datetime.now()` and `is_periodic_db_update_needed == True` 

Finally, we correctly start the actual polling again `dom_info_update_periodic_secs` after the previous poll cycle started.

I ran this again to confirm:

```
$ while true; do sonic-db-dump -n STATE_DB -y -k "TRANSCEIVER_DOM_SENSOR|Ethernet112" | grep last_update_time; sleep 10; done | uniq
      "last_update_time": "Tue Feb 24 19:17:09 2026",
      "last_update_time": "Tue Feb 24 19:18:10 2026",
      "last_update_time": "Tue Feb 24 19:19:10 2026",
      "last_update_time": "Tue Feb 24 19:20:10 2026",
      "last_update_time": "Tue Feb 24 19:21:11 2026",
      "last_update_time": "Tue Feb 24 19:22:11 2026",
```
There is a small drift - I attribute it to the 1 sec polling where - if the next polling cycle is to start in the next 1 to 1000msec, we will still wait the full 1000msec. Over time, the DOM update time will drift, but it is not drifting by one second every poll period.

It adds logic, but I can fix it by waiting the exact remaining time in the 1000msec loop if the amount of time left to start the next poll cycle is less than 1000msec

**Comment by mihirpat1** [comment]
@aditya-nexthop Can you please update the below HLD as well?
https://github.com/sonic-net/SONiC/blob/master/doc/platform_api/CMIS_Diagnostic_Monitoring_Overview_in_SONiC.md#5-sonic-cmis-diagnostic-monitoring-workflow

**Comment by aditya-nexthop** [comment]
> @aditya-nexthop Can you please update the below HLD as well? https://github.com/sonic-net/SONiC/blob/master/doc/platform_api/CMIS_Diagnostic_Monitoring_Overview_in_SONiC.md#5-sonic-cmis-diagnostic-monitoring-workflow

Done, please look at https://github.com/sonic-net/SONiC/pull/2303

File: sonic-xcvrd/xcvrd/dom/dom_mgr.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (mihirpat1) [reply sonic-xcvrd/xcvrd/dom/dom_mgr.py]: @aditya-nexthop Are you planning to fix this?

File: sonic-xcvrd/xcvrd/dom/dom_mgr.py [hunk 5]
```diff
@@ -277,42 +296,35 @@ def task_worker(self):
...
         while not self.task_stopping_event.is_set():
-            # Check if periodic db update is needed
-            now = datetime.datetime.now()
-            if next_periodic_db_update_time <= now:
-                is_periodic_db_update_needed = True
-
             # Handle port change event from main thread
             port_event_helper.handle_port_config_change(sel, asic_context, self.task_stopping_event, self.port_mapping, self.helper_logger, self.on_port_config_change)
 
+            stopping_event = False
```
Comment (mihirpat1) [inline sonic-xcvrd/xcvrd/dom/dom_mgr.py]: @aditya-nexthop Please rename `stopping_event` to `should_stop` or something meaningful.
The current name can cause confusion with `task_stopping_event`

Comment (aditya-nexthop) [reply sonic-xcvrd/xcvrd/dom/dom_mgr.py]: done

File: sonic-xcvrd/xcvrd/dom/dom_mgr.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (aditya-nexthop) [reply sonic-xcvrd/xcvrd/dom/dom_mgr.py]: fixed

**Review by mihirpat1 — COMMENTED** [review]
nit: the `should_stop` flag and the trailing `if should_stop: break` in the inner wait loop are redundant — the outer `while not self.task_stopping_event.is_set():` already re-checks the same condition and will exit on its own. Simplification:

```python
while True:
    remaining_secs = (next_periodic_db_update_time - datetime.datetime.now()).total_seconds()
    if remaining_secs <= 0:
        break
    select_timeout_msecs = min(PORT_UPDATE_EVENT_SELECT_TIMEOUT_MSECS,
                               max(1, int(remaining_secs * 1000)))
    self.check_port_update(port_change_observer, select_timeout_msecs)
    if self.task_stopping_event.is_set():
        self.log_notice("Stop event generated during DOM monitoring loop while checking port update")
        break
```

Same behavior (log message preserved), but the control flow is easier to follow.

**Review by mihirpat1 — DISMISSED** [review]

**Review by mihirpat1 — APPROVED** [review]

File: sonic-xcvrd/xcvrd/dom/dom_mgr.py [hunk 6]
```diff
@@ -277,42 +296,39 @@ def task_worker(self):
...
+            while True:
+                remaining_secs = (next_periodic_db_update_time - datetime.datetime.now()).total_seconds()
+                if remaining_secs <= 0:
+                    break
+                # Cap select timeout at PORT_UPDATE_EVENT_SELECT_TIMEOUT_MSECS to remain responsive
+                # to port change events, but shrink it to the remaining time when smaller so the
+                # next periodic DOM update is not delayed by an entire select cycle.
+                select_timeout_msecs = min(PORT_UPDATE_EVENT_SELECT_TIMEOUT_MSECS,
+                                           max(1, int(remaining_secs * 1000)))
+                self.check_port_update(port_change_observer, select_timeout_msecs)
```
Comment (prgeor) [inline sonic-xcvrd/xcvrd/dom/dom_mgr.py:314]: @aditya-nexthop wont this be simpler if we do
```
while True:
       remaining_secs = (next_periodic_db_update_time - datetime.datetime.now()).total_seconds()
       if remaining_secs * 1000 > PORT_UPDATE_EVENT_SELECT_TIMEOUT_MSECS 
             select_timeout_msecs = PORT_UPDATE_EVENT_SELECT_TIMEOUT_MSECS
             self.check_port_update(port_change_observer, select_timeout_msecs)
       else
             break;
```

Comment (prgeor) [reply sonic-xcvrd/xcvrd/dom/dom_mgr.py:314]: or you want to squeeze those <1000 msec into port update check?

Comment (aditya-nexthop) [reply sonic-xcvrd/xcvrd/dom/dom_mgr.py:314]: Yes we should, otherwise we cause a small drift over time as explained here.

https://github.com/sonic-net/sonic-platform-daemons/pull/758#discussion_r3042012608

**Review by prgeor — APPROVED** [review]
